dkg/state_machine.go

```diff
@@ -0,0 +1,659 @@
+var ErrCannotExecuteIfNotJoinerOrRemainer = errors.New("you cannot start execution if you are not a remainer or joiner to the DKG")
+var ErrUnknownAcceptor = errors.New("somebody unknown tried to accept the proposal")
+var ErrDuplicateAcceptance = errors.New("this participant already accepted the proposal")
+var ErrUnknownRejector = errors.New("somebody unknown tried to reject the proposal")
+var ErrDuplicateRejection = errors.New("this participant already rejected the proposal")
+var ErrNonLeaderCannotReceiveAcceptance = errors.New("you received an acceptance but are not the leader of this DKG - cannot do anything")
+var ErrNonLeaderCannotReceiveRejection = errors.New("you received a rejection but are not the leader of this DKG - cannot do anything")
+var ErrFinalGroupCannotBeEmpty = errors.New("you cannot complete a DKG with a nil final group")
+var ErrKeyShareCannotBeEmpty = errors.New("you cannot complete a DKG with a nil key share")
+
+//nolint:gocyclo
+func isValidStateChange(current, next DKGStatus) bool {
+	switch current {
+	case Complete:
+		return next == Proposing || next == Proposed
+	case Aborted:
+		return next == Proposing || next == Proposed
+	case TimedOut:
+		return next == Proposing || next == Proposed || next == Evicted
+	case Fresh:
+		return next == Proposing || next == Proposed
+	case Joined:
+		return next == Left || next == Executing || next == Aborted || next == TimedOut
+	case Left:
+		return next == Joined || next == Aborted
+	case Proposing:
+		return next == Executing || next == Aborted || next == TimedOut
+	case Proposed:
+		return next == Accepted || next == Rejected || next == Aborted || next == TimedOut || next == Left || next == Joined
+	case Accepted:
+		return next == Executing || next == Aborted || next == TimedOut
+	case Rejected:
+		return next == Aborted || next == TimedOut
+	case Executing:
+		return next == Complete || next == TimedOut || next == Evicted
+	case Evicted:
+		return next == Joined
+	}
+	return false
+}
+
+func hasTimedOut(details *DBState) bool {
+	now := time.Now()
+	return details.Timeout.Before(now) || details.Timeout.Equal(now)
+}
+
+func ValidateProposal(currentState *DBState, terms *drand.ProposalTerms) error {
+	if currentState.BeaconID != terms.BeaconID {
+		return ErrInvalidBeaconID
+	}
+
+	_, found := scheme.GetSchemeByID(terms.SchemeID)
+	if !found {
+		return ErrInvalidScheme
+	}
+
+	if terms.Timeout.AsTime().Before(time.Now()) {
+		return ErrTimeoutReached
+	}
+
+	if int(terms.Threshold) > len(terms.Joining)+len(terms.Remaining) {
+		return ErrThresholdHigherThanNodeCount
+	}
+
+	if err := validateEpoch(currentState, terms); err != nil {
+		return err
+	}
+
+	if terms.TransitionTime == nil {
+		return ErrTransitionTimeMissing
+	}
+
+	if terms.Epoch == 1 {
+		return validateFirstEpoch(terms)
+	}
+
+	// perhaps this should be stricter?
+	// should there be at least one round? should it be after `time.Now()`?
```

Comment:
Yes, to both, I think. Perhaps even a couple of rounds.
Thinking of higher frequency networks, e.g. 1 second ones, I'd say that this should take into account the network latency and perhaps be a fixed time in the future, e.g. 5 seconds?

Reply:
I think after `time.Now()` might be overly restrictive, as nodes could be receiving very delayed packets.
Network latency could be relevant, but nodes can catch up and aggregate old beacons okay; I don't know if I'm keen on a magic number fixed time, although I don't have an obvious reason why it shouldn't be

Reply:
Mhhhh TransitionTIme should be after `time.Now()` imo, because if it's not the case, that means they received the packet too late anyway and they won't do much, wont be able to participate in the DKG wont they ? 
If a leader sets transition time to genesis time + 1ms then the whole thing falls apart no ?